ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -58,12 +78,78 @@ func NewRayServiceReconciler(mgr manager.Manager) *RayServiceReconciler {
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2/pkg/reconcile
-func (r *RayServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = logf.FromContext(ctx)
+func (r *RayServiceReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("rayservice", request.NamespacedName)
+	log.Info("reconciling RayService", "service NamespacedName", request.NamespacedName)
+
+	// Get serving cluster instance
+	var err error
+	var rayServiceInstance *rayv1alpha1.RayService
+
+	if rayServiceInstance, err = r.getRayServiceInstance(request); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_TO_GET_RAYSERVICE, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var rayClusterInstance *rayv1alpha1.RayCluster
+	if rayClusterInstance, err = r.getOrCreateRayClusterInstance(rayServiceInstance); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_TO_GET_OR_CREATE_RAYCLUSTER, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	rayServiceInstance.Status.RayClusterStatus = rayClusterInstance.Status
+
+	var clientURL string
+	if clientURL, err = r.fetchDashboardURL(rayClusterInstance); err != nil || clientURL == "" {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.WAIT_FOR_DASHBOARD, err)
+		return ctrl.Result{}, err
+	}
+
+	rayDashboardClient := utils.RayDashboardClient{}
+	rayDashboardClient.InitClient(clientURL)
+
+	shouldUpdate := r.checkIfNeedSubmitServeDeployment(rayServiceInstance, request)
+
+	if shouldUpdate {
+		if err = r.updateServeDeployment(rayServiceInstance, rayDashboardClient, request); err != nil {
+			err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_SERVE_DEPLOY, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	var isHealthy bool
+	if isHealthy, err = r.getAndCheckServeStatus(rayServiceInstance, rayDashboardClient); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_GET_SERVE_DEPLOYMENT_STATUS, err)
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Check serve health", "isHealthy", isHealthy)
+
+	if isHealthy {
+		rayServiceInstance.Status.ServiceStatus = rayv1alpha1.RUNNING
+		if err := r.Status().Update(context.Background(), rayServiceInstance); err != nil {
```

Comment:
should we use the parent `ctx` from reconcile argumement?

Reply:
(How do contexts work here? rtfm is a valid answer :) )